Common/cpp/Tools/JsiUtils.h

```diff
@@ -0,0 +1,164 @@
+#pragma once
+
+#include <jsi/jsi.h>
+#include <string>
+#include <tuple>
+#include <utility>
+
+using namespace facebook;
+
+namespace reanimated {
+namespace jsi_utils {
+
+// `get` functions take a pointer to `jsi::Value` and
+// call an appropriate method to cast to the native type
+template <typename T>
+inline T get(jsi::Runtime &rt, const jsi::Value *value);
+
+template <>
+inline double get<double>(jsi::Runtime &rt, const jsi::Value *value) {
+  return value->asNumber();
+}
+
+template <>
+inline int get<int>(jsi::Runtime &rt, const jsi::Value *value) {
+  return value->asNumber();
+}
+
+template <>
+inline bool get<bool>(jsi::Runtime &rt, const jsi::Value *value) {
+  return value->asBool();
+}
+
+template <>
+inline jsi::Object get<jsi::Object>(jsi::Runtime &rt, const jsi::Value *value) {
+  return value->asObject(rt);
+}
+
+template <>
+inline jsi::Value const &get<jsi::Value const &>(
+    jsi::Runtime &rt,
+    const jsi::Value *value) {
+  return *value;
+}
+
+// `convertArgs` functions take a variadic template parameter of target (C++)
+// argument types `Targs` and a `jsi::Value` array `args`, and converts `args`
+// to a tuple of typed C++ arguments to be passed to the native implementation.
+// This is accomplished by dispatching (at compile time) to the correct
+// implementation based on the first type of `Targs`, using SFINAE to select the
+// correct specialization, and concatenating with the result of recursion on the
+// rest of `Targs`
+
+// BEGIN implementations for `convertArgs` specializations.
+// specialization for empty `Targs` - returns an empty tuple
+template <typename... Args>
+std::enable_if_t<(sizeof...(Args) == 0), std::tuple<>>
+convertArgs(jsi::Runtime &rt, const jsi::Value *args, const size_t count) {
+  assert(count == 0);
+  return std::make_tuple();
+}
+
+// calls `get<First>` on the first argument to retrieve the native type,
+// then calls recursively on the rest of `args`
+// and returns the concatenation of results
+template <typename T, typename... Rest>
+inline std::tuple<T, Rest...>
+convertArgs(jsi::Runtime &rt, const jsi::Value *args, const size_t count) {
+  auto arg = std::tuple<T>(get<T>(rt, args));
+  auto rest = convertArgs<Rest...>(rt, std::next(args), count - 1);
```

Comment:
Do we need to calculate and pass `count` around here? From what I can read we use `sizeof...(Args) == 0` to stop recursion and we use `count` just for assertion.

Reply:
It's to make sure the correct number of arguments are passed - if you pass too many, it will trip the assert in the `sizeof...(Args) == 0` case.
That said, a better place to do it would be in `getArgsForFunction`, to simply compare `sizeof...(Args)` with the `count` there.

Reply:
The reason I did it this way is, I used to accept the `Runtime` argument in arbitrary positions, in which case this function would add `rt` to the tuple and pass `args` and `count` unchanged.
Now I've decided to accept the `Runtime` only in the first argument to the host function, this is never the case.

Reply:
Also, my approach here _is_ wrong - if too few arguments were passed, it would happily read uninitialized memory here.
We have to check the `count` at the start (or check that it's >0 at every step).